src/main/java/com/google/devtools/build/lib/analysis/config/ConfigMatchingProvider.java

```diff
@@ -45,38 +45,63 @@ public abstract class ConfigMatchingProvider implements TransitiveInfoProvider {
    * <p>e.g. If merging where one is InError and the other is No, then currently will propagate the
    * errors, versus a more aggressive future approach could just propagate No.)
    */
-  // TODO(twigg): This is more cleanly implemented when Java record is available.
-  public static interface MatchResult {
-    // Only InError should return non-null.
-    public abstract @Nullable String getError();
-
-    public static final MatchResult MATCH = HasResult.MATCH;
-    public static final MatchResult NOMATCH = HasResult.NOMATCH;
-
-    /** Some specified error makes the match question irresolvable. */
-    @AutoValue
-    public abstract class InError implements MatchResult {
-      public static InError create(String error) {
-        return new AutoValue_ConfigMatchingProvider_MatchResult_InError(error);
+  public sealed interface MatchResult {
+    @AutoCodec
+    public record Match() implements MatchResult {}
+
+    MatchResult MATCH = new Match();
+
+    @AutoCodec
+    public record NoMatch(ImmutableList<Diff> diffs) implements MatchResult {
+      @AutoCodec.Instantiator
+      public NoMatch {}
+
+      public NoMatch(Diff diff) {
+        this(ImmutableList.of(diff));
       }
-    }
 
-    public static MatchResult create(boolean matches) {
-      return matches ? MATCH : NOMATCH;
+      @AutoCodec
+      public record Diff(Label what, String got, String want) {}
```

Comment:
I'm very worried about mixing up the want and got parameters. Any way this can be a builder style, so users can call `Diff.of(what).want(want).got(got).build()`? It's more verbose but less errorprone.